gcs/rewrite.py

```diff
@@ -56,25 +60,64 @@ def init_rest(
             },
             data=request.data,
         )
-        max_bytes_rewritten_per_call = min(
-            int(
-                fake_request.args.get("maxBytesRewrittenPerCall", cls.MIN_REWRITE_BYTES)
+        return cls(
+            request=request,
+            src_bucket_name=src_bucket_name,
+            src_object_name=src_object_name,
+            dst_bucket_name=dst_bucket_name,
+            dst_object_name=dst_object_name,
+            token=cls._token(),
+            media=b"",
+            max_bytes_rewritten_per_call=cls._normalize_max_bytes(
+                fake_request.args.get("maxBytesRewrittenPerCall")
             ),
-            cls.MIN_REWRITE_BYTES,
         )
-        token = hashlib.sha256(
-            (
-                "%s/o/%s/rewriteTo/b/%s/o/%s"
-                % (src_bucket_name, src_object_name, dst_bucket_name, dst_object_name)
-            ).encode("utf-8")
-        ).hexdigest()
+
+    @classmethod
+    def init_grpc(cls, request, context):
+        if not request.source_bucket.startswith("projects/_/buckets/"):
```

Comment:
Does JSON have this validation logic? If not should it be shared?

Reply:
JSON bucket names are not prefixed by `projects/_/buckets/` so the validation cannot be identical. I assume you are asking about the other checks (non-empty and so forth).  I think the bucket names being non-empty is validated implicitly by the Flask route decorators, e.g.:

https://github.com/googleapis/storage-testbench/blob/c1d3ec4e96a519e1b1f1d6077eaf96575678fb5c/testbench/rest_server.py#L595-L600

Reply:
Yep that's correct-- sounds good, thanks!